ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,520 @@
+	gt.balances[idx] = br
+	if 0 == atomic.AddInt64(&gt.incompleteCount, -1) {
+		groupDoneCh <- gt.groupTaskIndex
+	}
+}
+
+func (gt *groupTask) markCompletionResource(idx int, res loadedResourcesEntry, groupDoneCh chan int) {
+	gt.resources[idx] = res
+	if 0 == atomic.AddInt64(&gt.incompleteCount, -1) {
+		groupDoneCh <- gt.groupTaskIndex
+	}
+}
+
+func (gt *groupTask) markCompletionAcctError(err error, groupDoneCh chan int) {
+	for {
+		curVal := atomic.LoadInt64(&gt.incompleteCount)
+		if curVal <= 0 {
+			return
+		}
+		if atomic.CompareAndSwapInt64(&gt.incompleteCount, curVal, 0) {
+			gt.err = err
+			groupDoneCh <- -gt.groupTaskIndex
+			return
+		}
+	}
+}
+
+func (p *accountPrefetcher) asyncPrefetchRoutine(queue *preloaderTaskQueue, taskIdx *int64, groupDoneCh chan int) {
+	var task *preloaderTask
+	var err error
+	for {
+		nextTaskIdx := atomic.AddInt64(taskIdx, 1)
+		queue, task = queue.getTaskAtIndex(int(nextTaskIdx))
+		if task == nil {
+			// no more tasks.
+			return
+		}
+		if task.creatableIndex == 0 {
+			// lookup the account data directly from the ledger.
+			var acctData ledgercore.AccountData
+			acctData, _, err = p.ledger.LookupWithoutRewards(p.rnd, *task.address)
+			// if there was an error..
+			if err != nil {
+				// there was an error loading that entry.
+				break
+			}
+			br := loadedAccountDataEntry{
+				address: task.address,
+				data:    &acctData,
+			}
+			// update all the group tasks with the new acquired balance.
+			for i, wg := range task.groups {
+				wg.markCompletionAcct(task.groupIndices[i], br, groupDoneCh)
+			}
+			continue
+		}
+		if task.address == nil {
+			// start off by figuring out the creator in case it's a global resource.
+			var creator basics.Address
+			var ok bool
+			creator, ok, err = p.ledger.GetCreatorForRound(p.rnd, task.creatableIndex, task.creatableType)
+			if err != nil {
+				// there was an error loading that entry.
+				break
+			}
+			if !ok {
+				re := loadedResourcesEntry{
+					creatableIndex: task.creatableIndex,
+					creatableType:  task.creatableType,
+				}
+				// update all the group tasks with the new acquired balance.
+				for i, wg := range task.groups {
+					wg.markCompletionResource(task.groupIndices[i], re, groupDoneCh)
+				}
+				continue
+			}
+			task.address = &creator
+		}
+		var resource ledgercore.AccountResource
+		resource, err = p.ledger.LookupResource(p.rnd, *task.address, task.creatableIndex, task.creatableType)
+		if err != nil {
+			// there was an error loading that entry.
+			break
+		}
+		re := loadedResourcesEntry{
+			resource:       &resource,
+			address:        task.address,
+			creatableIndex: task.creatableIndex,
+			creatableType:  task.creatableType,
+		}
+		// update all the group tasks with the new acquired balance.
+		for i, wg := range task.groups {
+			wg.markCompletionResource(task.groupIndices[i], re, groupDoneCh)
+		}
+	}
+	// if we got here, it means that there was an error.
+	if task != nil {
```

Comment:
looks like it cannot be nil after `break`ing in the loop above

Reply:
that's true. I added the nil testing just to be on the safe side. This doesn't have any meaningful impact on performance since it would be executed at most once per go-routine.

Reply:
fine - I'm convinced. I'll add a comment about that.